src/supernova/snark.rs

```diff
@@ -0,0 +1,559 @@
+  C2: StepCircuit<G2::Scalar>,
+  S1: RelaxedR1CSSNARKTrait<G1>,
+  S2: RelaxedR1CSSNARKTrait<G2>,
+{
+  vks_primary: Vec<S1::VerifierKey>,
+  vk_secondary: S2::VerifierKey,
+  _p: PhantomData<(C1, C2)>,
+}
+
+/// The SNARK that proves the knowledge of a valid Supernova `RecursiveSNARK`
+pub struct CompressedSNARK<G1, G2, C1, C2, S1, S2>
+where
+  G1: Group<Base = <G2 as Group>::Scalar>,
+  G2: Group<Base = <G1 as Group>::Scalar>,
+  C1: StepCircuit<G1::Scalar>,
+  C2: StepCircuit<G2::Scalar>,
+  S1: RelaxedR1CSSNARKTrait<G1>,
+  S2: RelaxedR1CSSNARKTrait<G2>,
+{
+  r_U_primary: Vec<RelaxedR1CSInstance<G1>>,
+  r_W_snark_primary: Vec<S1>,
+
+  r_U_secondary: RelaxedR1CSInstance<G2>,
+  l_u_secondary: R1CSInstance<G2>,
+  nifs_secondary: NIFS<G2>,
+  f_W_snark_secondary: S2,
+
+  num_steps: usize,
+  program_counter: G1::Scalar,
+
+  zn_primary: Vec<G1::Scalar>,
+  zn_secondary: Vec<G2::Scalar>,
+
+  _p: PhantomData<(C1, C2)>,
+}
+
+impl<G1, G2, C1, C2, S1, S2> CompressedSNARK<G1, G2, C1, C2, S1, S2>
+where
+  G1: Group<Base = <G2 as Group>::Scalar>,
+  G2: Group<Base = <G1 as Group>::Scalar>,
+  C1: StepCircuit<G1::Scalar>,
+  C2: StepCircuit<G2::Scalar>,
+  S1: RelaxedR1CSSNARKTrait<G1>,
+  S2: RelaxedR1CSSNARKTrait<G2>,
+{
+  /// Generate the prover and verifier keys for the `CompressedSNARK` prover and verifier
+  pub fn setup(
+    pp: &PublicParams<G1, G2, C1, C2>,
+  ) -> Result<
+    (
+      ProverKey<G1, G2, C1, C2, S1, S2>,
+      VerifierKey<G1, G2, C1, C2, S1, S2>,
+    ),
+    SuperNovaError,
+  > {
+    let (pks_primary, vks_primary) = pp
+      .circuit_shapes
+      .iter()
+      .map(|c| S1::setup(&pp.ck_primary, &c.r1cs_shape))
+      .collect::<Result<Vec<_>, _>>()?
+      .into_iter()
+      .unzip();
+
+    let (pk_secondary, vk_secondary) =
+      S2::setup(&pp.ck_secondary, &pp.circuit_shape_secondary.r1cs_shape)?;
+
+    let prover_key = ProverKey {
+      pks_primary,
+      pk_secondary,
+      _p: PhantomData,
+    };
+    let verifier_key = VerifierKey {
+      vks_primary,
+      vk_secondary,
+      _p: PhantomData,
+    };
+
+    Ok((prover_key, verifier_key))
+  }
+
+  /// create a new `CompressedSNARK` proof
+  pub fn prove(
+    pp: &PublicParams<G1, G2, C1, C2>,
+    pk: &ProverKey<G1, G2, C1, C2, S1, S2>,
+    recursive_snark: &RecursiveSNARK<G1, G2>,
+  ) -> Result<Self, SuperNovaError> {
+    let res_secondary = NIFS::prove(
+      &pp.ck_secondary,
+      &pp.ro_consts_secondary,
+      &scalar_as_base::<G1>(pp.digest()),
+      &pp.circuit_shape_secondary.r1cs_shape,
+      recursive_snark.r_U_secondary[0].as_ref().unwrap(),
+      recursive_snark.r_W_secondary[0].as_ref().unwrap(),
+      &recursive_snark.l_u_secondary,
+      &recursive_snark.l_w_secondary,
+    );
+
+    let (nifs_secondary, (f_U_secondary, f_W_secondary)) = res_secondary?;
+
+    let r_W_snark_primary = pk
+      .pks_primary
+      .iter()
+      .zip(&recursive_snark.r_W_primary)
+      .zip(&recursive_snark.r_U_primary)
```

Comment:
I believe that this will produce a result that is shorter than `pk.pks_primary` if `recursive_snark.r_W_primary` or `recursive_snark.r_U_primary` are. That *could* be okay, since it will end up being an invalid `CompressedSNARK` as the result of trying to compress an invalid `RecursiveSNARK`. However, it's probably better to catch the problem and refuse to produce an invalid proof.

Note: it's extremely important that the *verifier* refuse to verify such invalid `CompressedSNARK`s — which I believe this initial implementation fails to do. I will comment below in more detail so you can verify.